rfcs/2022-10-31-13691-tooling-revamp.md

```diff
@@ -0,0 +1,209 @@
+
+```toml
+on = [
+  "src/sinks/elasticsearch",
+]
+args = [
+  "--features",
+  "es-integration-tests",
+  "--lib",
+  "::elasticsearch::integration_tests::",
+]
+
+[env]
+AWS_ACCESS_KEY_ID = "dummy"
+AWS_SECRET_ACCESS_KEY = "dummy"
+ELASTICSEARCH_AWS_ADDRESS = "http://localstack:4571"
+ELASTICSEARCH_HTTP_ADDRESS = "http://elasticsearch:9200"
+ELASTICSEARCH_HTTPS_ADDRESS = "https://elasticsearch-secure:9200"
+
+[[matrix]]
+version = ["7.13.1"]
+```
+
+We could then extend the support matrices:
+
+```toml
+[[matrix]]
+version = ["7.13.1", "8.4.3"]
+type = ["classic"]
+
+[[matrix]]
+version = ["1.3.6", "2.3.0"]
+type = ["opensearch"]
+```
+
+That would indicate the following unique environments:
+
+```text
+7.13.1-classic
+8.4.3-classic
+1.3.6-opensearch
+2.3.0-opensearch
+```
+
+If we were testing the `8.4.3-classic` environment the [management CLI](#environment-management) would get the following payload:
+
+```json
+{
+    "type": "classic",
+    "version": "8.4.3"
+}
+```
+
+#### Interface
+
+The `vdev int` command would have the following sub-commands:
+
+- `show` - accepts an integration name and shows the available/running environments
+- `start` - calls the equivalently named command of the [management CLI](#environment-management). requires an integration name and environment name. the integration names refer to the names of the integration test directories
+- `stop` - ^^
+- `test` - executes the tests and will terminate with the same exit code. accepts an integration name, environment name, and arbitrary extra arguments to pass to the test command. if no environment name is provided then all will be tested and any not already started will be torn down
+
+The [testing container](#testing-container) will be created first always.
+
+#### Testing container
+
+There will be a single container per integration for running tests. The entry point will be `/bin/sleep infinity` and will be attached to the host network.
+
+#### CI
+
+The test matrix on PRs will be computed ([example](https://github.com/vectordotdev/vector/blob/v0.24.2/.github/workflows/soak.yml#L137-L169)) by `vdev` based on what has changed. This will save [somewhere between](https://github.com/vectordotdev/vector/actions/workflows/integration-test.yml) 6.5-8 hours of billable CI time per commit.
+
+## Drawbacks
+
+- Slight change in process for all developers which would require communication and docs
+
+## Prior Art
+
+- [ddev](https://datadoghq.dev/integrations-core/ddev/cli/) is a tool specific to Datadog Agent integrations
+- [Hatch](https://github.com/pypa/hatch) provides environment management and the matrix syntax we borrowed
+
+## Alternatives
+
+### Status quo
+
+- Could ignore all pain and add some more Docker compose test files
+
+### Python
+
+- Could call environments' `start`/`stop` methods through Hatch to avoid maintaining matrix logic
+- Could write all tooling in Python to theoretically increase development velocity
```

Comment:
I also thought of this yesterday - if this is to power CI, how/when would it be built for CI?

If we need to compiler during the the CI run that adds time (hopefully not much because the code should be small...) or built only on changes to the tool and pull from those artifacts.

Reply:
> As another data point, @tobz how would you have felt writing `generate-components-docs.rb` in Rust vs. Ruby?

At the time, it probably would have sucked more to slog through writing it in Rust from scratch. Doing that _now_... probably wouldn't be terrible because it'd be easy to know when I reached feature parity by comparing the outputs directly, and having a good mental model of how it's structured.

🤷🏻